File: 2015/go/03.go
package main

import (
	"fmt"
	"io/ioutil"
)

type vec struct {
	x int
	y int
}

func contains(s []vec, e vec) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}

func answer1(input string) int {
	santa := vec{0, 0}
	visited := []vec{santa}
	for _, char := range input {
		c := string(char)
		switch c {
		case "^":
			santa.y++
		case "v":
			santa.y--
		case ">":
			santa.x++
		case "<":
			santa.x--
		}
		value := vec{santa.x, santa.y}
		if !contains(visited, value) {
			visited = append(visited, value)
		}
	}
	return len(visited)
}

func answer2(input string) int {
	santa := vec{0, 0}
	robot := vec{0, 0}
	turn := true
	visited := []vec{santa}
	for _, char := range input {
		c := string(char)
		var actual *vec
		if turn {
			actual = &santa
		} else {
			actual = &robot
		}
		switch c {
		case "^":
			actual.y++
		case "v":
			actual.y--
		case ">":
			actual.x++
		case "<":
			actual.x--
		}
		if !contains(visited, *actual) {
			visited = append(visited, *actual)
		}
		turn = !turn
	}
	return len(visited)
}

func main() {
	data, _ := ioutil.ReadFile("03input.txt")
	text := string(data)
	fmt.Println(answer1(text))
	fmt.Println(answer2(text))
}

File: 2015/03.lua
local aoc = require "lib.aoc"
local input = aoc.input.getInput()

local function answer1()
	local visits = {}
	local santa = { x = 0, y = 0 }
	local totalVisits = 1
	local key = santa.x .. "," .. santa.y
	visits[key] = true

	for _, char in aoc.string.iterate(input) do
		if char == ">" then
			santa.x = santa.x + 1
		elseif char == "^" then
			santa.y = santa.y - 1
		elseif char == "<" then
			santa.x = santa.x - 1
		elseif char == "v" then
			santa.y = santa.y + 1
		end

		key = santa.x .. "," .. santa.y

		if visits[key] == nil then
			visits[key] = true
			totalVisits = totalVisits + 1
		end
	end
	return totalVisits
end

local function answer2()
	local visits = {}
	local santa = { x = 0, y = 0 }
	local roboSanta = { x = 0, y = 0 }
	local totalVisits = 1
	local key = santa.x .. "," .. santa.y
	local turn = true
	visits[key] = true

	for _, char in aoc.string.iterate(input) do
		local whoToMove = turn and santa or roboSanta

		if char == ">" then
			whoToMove.x = whoToMove.x + 1
		elseif char == "^" then
			whoToMove.y = whoToMove.y - 1
		elseif char == "<" then
			whoToMove.x = whoToMove.x - 1
		elseif char == "v" then
			whoToMove.y = whoToMove.y + 1
		end

		key = whoToMove.x .. "," .. whoToMove.y

		if visits[key] == nil then
			visits[key] = true
			totalVisits = totalVisits + 1
		end

		turn = not turn
	end
	return totalVisits
end
print("answer 1 is " .. answer1())
print("answer 2 is " .. answer2())
